### PR TITLE
[BlenderBot2] clean up reply during interactive mode

### PR DESCRIPTION
**Pull request by jxmsML**
**Patch description**
Right now, during interactive mode, the `'_POTENTIALLY_UNSAFE__'` from blenderbot2 generation would go into blenderbot 2 history and possibly affects its subsequent generation. However the model is never trained on observing such specially tokens, only generating them.

After the change, we still keep the `'_POTENTIALLY_UNSAFE__'`  in the reply text, and only remove that from the blenderbot2 history.

**Testing steps**

**Other information**

### CODE REVIEW

File: projects/blenderbot2/agents/blenderbot2.py [hunk 1]
```diff
@@ -898,6 +905,39 @@ def compute_loss(
...
+        """
+        Observe one's own utterance.
+        Override TorchAgent.self_observe with the optional cleaned text
+
+        :param self_message:
+            The message corresponding to the output from batch_act.
+        """
+
+        if (
+            self.opt('clean_reply', False)
```
Comment (mojtaba-komeili) [inline projects/blenderbot2/agents/blenderbot2.py:922]: So it returns here and doesn't go to cleaning the text if `clean_reply` is `True`?

Comment (jxmsML) [reply projects/blenderbot2/agents/blenderbot2.py:922]: I'd love to ALWAYS clean the reply and I think it's the right thing to do?

File: projects/blenderbot2/agents/blenderbot2.py [hunk 2]
```diff
@@ -898,6 +905,39 @@ def compute_loss(
...
+            or self.observation['episode_done']  # last example in the episode
+            or use_reply == 'none'  # not including our own responses anyway
```
Comment (mojtaba-komeili) [inline projects/blenderbot2/agents/blenderbot2.py:924]: Where does `use_reply` come from?

File: projects/blenderbot2/agents/blenderbot2.py (hunk 1 above)
Comment (jxmsML) [reply projects/blenderbot2/agents/blenderbot2.py]: I'd love to ALWAYS clean the reply and I think it's the right thing to do?

File: projects/blenderbot2/agents/blenderbot2.py (hunk 2 above)
Comment (jxmsML) [reply projects/blenderbot2/agents/blenderbot2.py:918]: good catch, I've added that back now.

**Review by klshuster — APPROVED** [review]

**Comment by klshuster** [comment]
could you maybe add some test cases to make sure that `_clean_text` works?

File: projects/blenderbot2/agents/blenderbot2.py [hunk 3]
```diff
@@ -898,6 +899,38 @@ def compute_loss(
...
+                and any([l in self.observation for l in ['labels', 'eval_labels']])
+            )  # has true label
+        ):
+            return super().self_observe(self_message)
+
+        # otherwise, we use the CLEANED last output the model generated
+        if self_message is not None:
+            last_reply = self_message['text']
+            clean_reply = self._clean_text(last_reply)
+            self.history.add_reply(clean_reply)
```
Comment (stephenroller) [inline projects/blenderbot2/agents/blenderbot2.py:929]: perhaps we should just override the History class and do this logic there?

**Comment by stephenroller** [comment]
Go ahead and merge main into this branch

**Pull request merged**